### PR TITLE
Fix a logic error in eraseUseOfValue.

### PR DESCRIPTION
**Pull request by trentxintong**

### CODE REVIEW

**Comment by trentxintong** [comment]
@swift-ci Please test

**Comment by trentxintong** [comment]
@swift-ci Please ASAN test

File: lib/SILOptimizer/Utils/Local.cpp [hunk 1]
```diff
@@ -208,6 +208,10 @@ void swift::eraseUsesOfValue(SILValue V) {
     auto *User = UI->getUser();
     UI++;
 
+    // User could be removed at this point.
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Local.cpp]: Is this the right fix? If the User is removed are we guaranteed to have it be nulled out?

Why not use a SmallPointerSet.

Comment (trentxintong) [inline lib/SILOptimizer/Utils/Local.cpp]: One possibility is to collect all the direct and transitive users of the SILValue and remove them one by one.  But that may leave us with some dead, but not removed instructions.

**Comment by trentxintong** [comment]
@swift-ci Please test Linux

**Comment by trentxintong** [comment]
@swift-ci Please test

**Comment by trentxintong** [comment]
@swift-ci Please test

**Pull request merged**